src/Domain/CoordinateMaps/KerrHorizon.cpp

```diff
@@ -0,0 +1,177 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Domain/CoordinateMaps/KerrHorizon.hpp"
+#include "DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/DereferenceWrapper.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeArray.hpp"
+#include "Utilities/StdArrayHelpers.hpp"
+
+namespace domain::CoordinateMaps {
+
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> KerrHorizon::operator()(
+    const std::array<T, 3>& source_coords) const noexcept {
+  return make_array<tt::remove_cvref_wrap_t<T>, 3>(source_coords) *
+         sqrt(stretch_factor(source_coords));
+}
+
+std::optional<std::array<double, 3>> KerrHorizon::inverse(
+    const std::array<double, 3>& target_coords) const noexcept {
+  const auto coords_mag_sq = dot(target_coords, target_coords);
+  const auto coords_sq_min_spin_sq = coords_mag_sq - spin_mag_sq_;
+  const auto coords_dot_spin = dot(target_coords, spin_);
+  const auto fac = (coords_sq_min_spin_sq +
+                    sqrt(coords_sq_min_spin_sq * coords_sq_min_spin_sq +
+                         4. * square(coords_dot_spin))) /
+                   (2. * coords_mag_sq);
+  // this is the way it was written in spec, but I dont think `fac` can
+  // ever be smaller than 0
+  return fac >= 0.
+             ? std::optional<std::array<double, 3>>(target_coords * sqrt(fac))
+             : std::nullopt;
+}
+
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame> KerrHorizon::jacobian(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+
+  tnsr::Ij<ReturnType, 3, Frame::NoFrame> jac(
+      get_size(dereference_wrapper(source_coords[0])));
+
+  // use allocations from `jac` for auxiliaries
+  ReturnType& fac = get<0, 0>(jac);
+  ReturnType& source_coords_sq = get<0, 1>(jac);
+  ReturnType& coords_dot_spin = get<0, 2>(jac);
+  ReturnType& subexpr_1 = get<1, 0>(jac);
+  ReturnType& subexpr_2 = get<1, 1>(jac);
+
+  std::array<ReturnType, 3> dfac_dx{};
+  if constexpr (std::is_same_v<ReturnType, DataVector>) {
+    dfac_dx[0].set_data_ref(&get<2, 0>(jac));
+    dfac_dx[1].set_data_ref(&get<2, 1>(jac));
+    dfac_dx[2].set_data_ref(&get<2, 2>(jac));
+  }
+
+  fac = stretch_factor(source_coords);
+  source_coords_sq = dot(source_coords, source_coords);
+  coords_dot_spin = dot(source_coords, spin_);
+  subexpr_1 = 4. * source_coords_sq * (1. - fac);
+  subexpr_2 = 2. * fac * coords_dot_spin;
+
+  for (size_t i = 0; i < 3; ++i) {
+    gsl::at(dfac_dx, i) = subexpr_1 * gsl::at(source_coords, i) +
+                          2. * spin_mag_sq_ * gsl::at(source_coords, i) -
+                          subexpr_2 * gsl::at(spin_, i);
+  }
+  dfac_dx = dfac_dx / (square(source_coords_sq) + square(coords_dot_spin));
+
+  const ReturnType sqrt_fac = sqrt(fac);
```

Comment:
could you use the `<2,2>` component to get rid of this allocation?

Reply:
Im already using that component for  `dfac_dx[2]`.